src/backends/github/API.js

```diff
@@ -507,42 +458,35 @@ export default class API {
         content: contentBase64,
         encoding: "base64",
       }),
-    }).then((response) => {
-      item.sha = response.sha;
-      item.uploaded = true;
-      return item;
+    })).then(response => Object.assign({}, item, {
+      sha: response.sha,
+      uploaded: true,
     }));
   }
 
+  isFile(obj) {
+    return obj.file;
+  }
+
   updateTree(sha, path, fileTree) {
     return this.getTree(sha)
-      .then((tree) => {
-        let obj;
-        let filename;
-        let fileOrDir;
-        const updates = [];
-        const added = {};
-
-        for (let i = 0, len = tree.tree.length; i < len; i++) {
-          obj = tree.tree[i];
-          if (fileOrDir = fileTree[obj.path]) {
-            added[obj.path] = true;
-            if (fileOrDir.file) {
-              updates.push({ path: obj.path, mode: obj.mode, type: obj.type, sha: fileOrDir.sha });
-            } else {
-              updates.push(this.updateTree(obj.sha, obj.path, fileOrDir));
-            }
-          }
-        }
-        for (filename in fileTree) {
-          fileOrDir = fileTree[filename];
-          if (added[filename]) { continue; }
-          updates.push(
-            fileOrDir.file ?
-              { path: filename, mode: "100644", type: "blob", sha: fileOrDir.sha } :
-              this.updateTree(null, filename, fileOrDir)
-          );
-        }
+      .then(({ tree: dirContents }) => {
+        const updatedItems = dirContents.filter(item => fileTree[item.path]);
+        const added = zipObject(updatedItems.map(item => item.path), fill([], true, 0, updatedItems.length));
+        const [updatedFiles, updatedDirs] = partition(updatedItems, this.isFile);
+        const updatePromises = [
+          ...updatedDirs.map(dir => this.updateTree(dir.sha, dir.path, fileTree[dir.path])),
+          ...updatedFiles.map(file => ({ ...pick(file, ['path', 'mode', 'type']), sha: fileTree[file.path].sha })),
+        ];
+
+        const newItems = Object.entries(fileTree).filter(([filename]) => !added[filename]);
+        const [newFiles, newDirs] = partition(newItems, ([, file]) => this.isFile(file));
```

Comment:
@tortilaman it's destructuring the argument and ignoring the first element. For example:

```
// these are equivalent

const f1 = ([ x, y ]) => y > 0
const f2 = ([, y]) => y > 0

f1(0, 1) // true
f2(0, 1) // true
f1(1, 0) // false
f2(1, 0) // false
```

It allows us to prevent creating a name for something we never use later on. If it's obscure enough it's confusing to most people, I can change it back.